File: lib/tauri/calculator.ts
/**
 * Calculator operations with Tauri/browser fallback
 */

import { isTauri, invoke } from './platform';
import type { SimpleSequence, Coordinates } from '../nina/simple-sequence-types';

export interface RaResult {
  hours: number;
  minutes: number;
  seconds: number;
}

export interface DecResult {
  degrees: number;
  minutes: number;
  seconds: number;
  negative: boolean;
}

/**
 * Calculate sequence runtime
 */
export async function calculateSequenceRuntime(sequence: SimpleSequence): Promise<number> {
  if (isTauri()) {
    return invoke<number>('calculate_sequence_runtime', { sequence });
  }
  
  // Browser fallback
  let total = 0;
  for (const target of sequence.targets) {
    total += target.delay;
    for (const exp of target.exposures) {
      if (exp.enabled) {
        const remaining = Math.max(0, exp.totalCount - exp.progressCount);
        total += remaining * (exp.exposureTime + sequence.estimatedDownloadTime);
      }
    }
  }
  return total;
}

/**
 * Calculate sequence ETAs
 */
export async function calculateSequenceEtas(sequence: SimpleSequence): Promise<SimpleSequence> {
  if (isTauri()) {
    return invoke<SimpleSequence>('calculate_sequence_etas', { sequence });
  }
  
  // Browser fallback
  const result = { ...sequence };
  const downloadTime = sequence.estimatedDownloadTime;
  let currentTime = new Date();
  let totalDuration = 0;
  
  result.targets = sequence.targets.map(target => {
    const newTarget = { ...target };
    let targetDuration = target.delay;
    
    for (const exp of target.exposures) {
      if (exp.enabled) {
        const remaining = Math.max(0, exp.totalCount - exp.progressCount);
        targetDuration += remaining * (exp.exposureTime + downloadTime);
      }
    }
    
    newTarget.estimatedStartTime = new Date(currentTime);
    newTarget.estimatedDuration = targetDuration;
    currentTime = new Date(currentTime.getTime() + targetDuration * 1000);
    newTarget.estimatedEndTime = new Date(currentTime);
    totalDuration += targetDuration;
    
    return newTarget;
  });
  
  result.overallStartTime = new Date();
  result.overallEndTime = new Date(Date.now() + totalDuration * 1000);
  result.overallDuration = totalDuration;
  
  return result;
}

/**
 * Format duration
 */
export async function formatDuration(seconds: number): Promise<string> {
  if (isTauri()) {
    return invoke<string>('format_duration', { seconds });
  }
  
  // Browser fallback
  if (seconds < 0) return '0s';
  
  const days = Math.floor(seconds / 86400);
  const hours = Math.floor((seconds % 86400) / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);
  const secs = Math.floor(seconds % 60);
  
  if (days > 0) {
    return `${days}d ${hours}h ${minutes}m ${secs}s`;
  } else if (hours > 0) {
    return `${hours}h ${minutes}m ${secs}s`;
  } else if (minutes > 0) {
    return `${minutes}m ${secs}s`;
  } else {
    return `${secs}s`;
  }
}

/**
 * Calculate angular separation
 */
export async function calculateAngularSeparation(
  coord1: Coordinates,
  coord2: Coordinates
): Promise<number> {
  if (isTauri()) {
    return invoke<number>('calculate_angular_separation', { coord1, coord2 });
  }
  
  // Browser fallback
  const ra1 = (coord1.raHours + coord1.raMinutes / 60 + coord1.raSeconds / 3600) * 15 * Math.PI / 180;
  const dec1 = (coord1.decDegrees + coord1.decMinutes / 60 + coord1.decSeconds / 3600) * (coord1.negativeDec ? -1 : 1) * Math.PI / 180;
  const ra2 = (coord2.raHours + coord2.raMinutes / 60 + coord2.raSeconds / 3600) * 15 * Math.PI / 180;
  const dec2 = (coord2.decDegrees + coord2.decMinutes / 60 + coord2.decSeconds / 3600) * (coord2.negativeDec ? -1 : 1) * Math.PI / 180;
  
  const deltaRa = ra2 - ra1;
  const cosSep = Math.sin(dec1) * Math.sin(dec2) + Math.cos(dec1) * Math.cos(dec2) * Math.cos(deltaRa);
  
  return Math.acos(Math.min(1, Math.max(-1, cosSep))) * 180 / Math.PI;
}

/**
 * Convert RA to decimal
 */
export async function raToDecimal(hours: number, minutes: number, seconds: number): Promise<number> {
  if (isTauri()) {
    return invoke<number>('ra_to_decimal', { hours, minutes, seconds });
  }
  return hours + minutes / 60 + seconds / 3600;
}

/**
 * Convert decimal to RA
 */
export async function decimalToRa(decimal: number): Promise<RaResult> {
  if (isTauri()) {
    return invoke<RaResult>('decimal_to_ra', { decimal });
  }
  
  const hours = Math.floor(decimal);
  const minutesDecimal = (decimal - hours) * 60;
  const minutes = Math.floor(minutesDecimal);
  const seconds = Math.round((minutesDecimal - minutes) * 60 * 100) / 100;
  
  return { hours, minutes, seconds };
}

/**
 * Convert Dec to decimal
 */
export async function decToDecimal(
  degrees: number,
  minutes: number,
  seconds: number,
  negative: boolean
): Promise<number> {
  if (isTauri()) {
    return invoke<number>('dec_to_decimal', { degrees, minutes, seconds, negative });
  }
  
  const value = Math.abs(degrees) + minutes / 60 + seconds / 3600;
  return negative ? -value : value;
}

/**
 * Convert decimal to Dec
 */
export async function decimalToDec(decimal: number): Promise<DecResult> {
  if (isTauri()) {
    return invoke<DecResult>('decimal_to_dec', { decimal });
  }
  
  const negative = decimal < 0;
  const absDecimal = Math.abs(decimal);
  const degrees = Math.floor(absDecimal);
  const minutesDecimal = (absDecimal - degrees) * 60;
  const minutes = Math.floor(minutesDecimal);
  const seconds = Math.round((minutesDecimal - minutes) * 60 * 100) / 100;
  
  return { degrees, minutes, seconds, negative };
}

/**
 * Calculate altitude
 */
export async function calculateAltitude(
  raHours: number,
  decDegrees: number,
  latitude: number,
  longitude: number,
  datetime?: string
): Promise<number> {
  if (isTauri()) {
    return invoke<number>('calculate_altitude', { raHours, decDegrees, latitude, longitude, datetime });
  }
  
  // Browser fallback - simplified calculation
  // Note: This is a rough approximation assuming object is at meridian
  // For accurate calculations, use the Tauri backend
  void longitude; // unused in simplified calculation
  const altitude = 90 - Math.abs(latitude - decDegrees);
  return Math.max(-90, Math.min(90, altitude));
}

/**
 * Check if object is above horizon
 */
export async function isAboveHorizon(
  raHours: number,
  decDegrees: number,
  latitude: number,
  longitude: number,
  minAltitude: number,
  datetime?: string
): Promise<boolean> {
  const altitude = await calculateAltitude(raHours, decDegrees, latitude, longitude, datetime);
  return altitude >= minAltitude;
}

/**
 * Calculate moon phase
 */
export async function calculateMoonPhase(datetime?: string): Promise<number> {
  if (isTauri()) {
    return invoke<number>('calculate_moon_phase', { datetime });
  }
  
  // Browser fallback - simplified calculation
  const date = datetime ? new Date(datetime) : new Date();
  const year = date.getFullYear();
  const month = date.getMonth() + 1;
  const day = date.getDate();
  
  // Simplified moon phase calculation
  const c = Math.floor(365.25 * year);
  const e = Math.floor(30.6 * month);
  const jd = c + e + day - 694039.09;
  const phase = jd / 29.530588853;
  
  return phase - Math.floor(phase);
}

/**
 * Calculate moon illumination
 */
export async function calculateMoonIllumination(datetime?: string): Promise<number> {
  if (isTauri()) {
    return invoke<number>('calculate_moon_illumination', { datetime });
  }
  
  const phase = await calculateMoonPhase(datetime);
  const angle = phase * 2 * Math.PI;
  return (1 - Math.cos(angle)) / 2 * 100;
}

/**
 * Parse RA string
 */
export async function parseRa(raString: string): Promise<RaResult | null> {
  if (isTauri()) {
    try {
      return await invoke<RaResult>('parse_ra', { raString });
    } catch {
      return null;
    }
  }
  
  // Browser fallback
  const match = raString.match(/(\d+)[h:\s]+(\d+)[m:\s]+(\d+\.?\d*)/);
  if (!match) return null;
  
  return {
    hours: parseInt(match[1], 10),
    minutes: parseInt(match[2], 10),
    seconds: parseFloat(match[3]),
  };
}

/**
 * Parse Dec string
 */
export async function parseDec(decString: string): Promise<DecResult | null> {
  if (isTauri()) {
    try {
      return await invoke<DecResult>('parse_dec', { decString });
    } catch {
      return null;
    }
  }
  
  // Browser fallback
  const match = decString.match(/([+-]?)(\d+)[°:\s]+(\d+)[':\s]+(\d+\.?\d*)/);
  if (!match) return null;
  
  return {
    negative: match[1] === '-',
    degrees: parseInt(match[2], 10),
    minutes: parseInt(match[3], 10),
    seconds: parseFloat(match[4]),
  };
}

/**
 * Format RA
 */
export async function formatRa(hours: number, minutes: number, seconds: number): Promise<string> {
  if (isTauri()) {
    return invoke<string>('format_ra', { hours, minutes, seconds });
  }
  
  return `${hours.toString().padStart(2, '0')}h ${minutes.toString().padStart(2, '0')}m ${seconds.toFixed(1)}s`;
}

/**
 * Format Dec
 */
export async function formatDec(
  degrees: number,
  minutes: number,
  seconds: number,
  negative: boolean
): Promise<string> {
  if (isTauri()) {
    return invoke<string>('format_dec', { degrees, minutes, seconds, negative });
  }
  
  const sign = negative ? '-' : '+';
  return `${sign}${degrees}° ${minutes.toString().padStart(2, '0')}' ${seconds.toFixed(1)}"`;
}

File: src-tauri/src/services/calculator.rs
//! Calculation services for astronomy and sequence timing

use crate::models::*;
use chrono::{DateTime, Utc, Duration};

/// Calculate total runtime for a simple sequence
pub fn calculate_sequence_runtime(sequence: &SimpleSequence) -> f64 {
    sequence.total_runtime()
}

/// Calculate ETA for all targets in a sequence
pub fn calculate_sequence_etas(sequence: &mut SimpleSequence) {
    sequence.calculate_etas();
}

/// Calculate exposure runtime
pub fn calculate_exposure_runtime(exposure: &SimpleExposure, download_time: f64) -> f64 {
    exposure.runtime(download_time)
}

/// Calculate target runtime
pub fn calculate_target_runtime(target: &SimpleTarget, download_time: f64) -> f64 {
    target.runtime(download_time)
}

/// Format duration in human-readable format
pub fn format_duration(seconds: f64) -> String {
    if seconds < 0.0 {
        return "0s".to_string();
    }
    
    let total_seconds = seconds as i64;
    let days = total_seconds / 86400;
    let hours = (total_seconds % 86400) / 3600;
    let minutes = (total_seconds % 3600) / 60;
    let secs = total_seconds % 60;
    
    if days > 0 {
        format!("{}d {}h {}m {}s", days, hours, minutes, secs)
    } else if hours > 0 {
        format!("{}h {}m {}s", hours, minutes, secs)
    } else if minutes > 0 {
        format!("{}m {}s", minutes, secs)
    } else {
        format!("{}s", secs)
    }
}

/// Format time as HH:MM:SS
pub fn format_time(datetime: DateTime<Utc>) -> String {
    datetime.format("%H:%M:%S").to_string()
}

/// Calculate end time from start time and duration
pub fn calculate_end_time(start: DateTime<Utc>, duration_seconds: f64) -> DateTime<Utc> {
    start + Duration::seconds(duration_seconds as i64)
}

/// Angular separation between two coordinates in degrees
pub fn angular_separation(coord1: &Coordinates, coord2: &Coordinates) -> f64 {
    coordinates::angular_separation(coord1, coord2)
}

/// Convert RA from HMS to decimal hours
pub fn ra_to_decimal(hours: i32, minutes: i32, seconds: f64) -> f64 {
    hours as f64 + minutes as f64 / 60.0 + seconds / 3600.0
}

/// Convert RA from decimal hours to HMS
pub fn decimal_to_ra(decimal: f64) -> (i32, i32, f64) {
    let hours = decimal.floor() as i32;
    let minutes_decimal = (decimal - hours as f64) * 60.0;
    let minutes = minutes_decimal.floor() as i32;
    let seconds = (minutes_decimal - minutes as f64) * 60.0;
    (hours, minutes, (seconds * 100.0).round() / 100.0)
}

/// Convert Dec from DMS to decimal degrees
pub fn dec_to_decimal(degrees: i32, minutes: i32, seconds: f64, negative: bool) -> f64 {
    let value = degrees.abs() as f64 + minutes as f64 / 60.0 + seconds / 3600.0;
    if negative { -value } else { value }
}

/// Convert Dec from decimal degrees to DMS
pub fn decimal_to_dec(decimal: f64) -> (i32, i32, f64, bool) {
    let negative = decimal < 0.0;
    let abs_decimal = decimal.abs();
    let degrees = abs_decimal.floor() as i32;
    let minutes_decimal = (abs_decimal - degrees as f64) * 60.0;
    let minutes = minutes_decimal.floor() as i32;
    let seconds = (minutes_decimal - minutes as f64) * 60.0;
    (degrees, minutes, (seconds * 100.0).round() / 100.0, negative)
}

/// Calculate altitude of an object at a given time
/// This is a simplified calculation - for accurate results, use a proper astronomy library
pub fn calculate_altitude(
    ra_hours: f64,
    dec_degrees: f64,
    latitude: f64,
    longitude: f64,
    datetime: DateTime<Utc>,
) -> f64 {
    // Convert to radians
    let lat_rad = latitude.to_radians();
    let dec_rad = dec_degrees.to_radians();
    
    // Calculate Local Sidereal Time (simplified)
    let jd = datetime_to_julian_day(datetime);
    let lst = calculate_lst(jd, longitude);
    
    // Calculate Hour Angle
    let ha = (lst - ra_hours * 15.0).to_radians();
    
    // Calculate altitude
    let sin_alt = lat_rad.sin() * dec_rad.sin() + lat_rad.cos() * dec_rad.cos() * ha.cos();
    sin_alt.asin().to_degrees()
}

/// Convert DateTime to Julian Day
fn datetime_to_julian_day(datetime: DateTime<Utc>) -> f64 {
    let year = datetime.format("%Y").to_string().parse::<i32>().unwrap();
    let month = datetime.format("%m").to_string().parse::<i32>().unwrap();
    let day = datetime.format("%d").to_string().parse::<f64>().unwrap();
    let hour = datetime.format("%H").to_string().parse::<f64>().unwrap();
    let minute = datetime.format("%M").to_string().parse::<f64>().unwrap();
    let second = datetime.format("%S").to_string().parse::<f64>().unwrap();
    
    let day_fraction = day + (hour + minute / 60.0 + second / 3600.0) / 24.0;
    
    let (y, m) = if month <= 2 {
        (year - 1, month + 12)
    } else {
        (year, month)
    };
    
    let a = (y as f64 / 100.0).floor();
    let b = 2.0 - a + (a / 4.0).floor();
    
    (365.25 * (y as f64 + 4716.0)).floor() 
        + (30.6001 * (m as f64 + 1.0)).floor() 
        + day_fraction + b - 1524.5
}

/// Calculate Local Sidereal Time in degrees
fn calculate_lst(jd: f64, longitude: f64) -> f64 {
    let t = (jd - 2451545.0) / 36525.0;
    let gmst = 280.46061837 
        + 360.98564736629 * (jd - 2451545.0) 
        + 0.000387933 * t * t 
        - t * t * t / 38710000.0;
    
    let lst = gmst + longitude;
    lst.rem_euclid(360.0)
}

/// Estimate if an object is above horizon
pub fn is_above_horizon(
    ra_hours: f64,
    dec_degrees: f64,
    latitude: f64,
    longitude: f64,
    datetime: DateTime<Utc>,
    min_altitude: f64,
) -> bool {
    calculate_altitude(ra_hours, dec_degrees, latitude, longitude, datetime) >= min_altitude
}

/// Calculate moon phase (0 = new, 0.5 = full, 1 = new)
pub fn calculate_moon_phase(datetime: DateTime<Utc>) -> f64 {
    let jd = datetime_to_julian_day(datetime);
    let days_since_new = (jd - 2451550.1) % 29.530588853;
    days_since_new / 29.530588853
}

/// Calculate moon illumination percentage
pub fn calculate_moon_illumination(datetime: DateTime<Utc>) -> f64 {
    let phase = calculate_moon_phase(datetime);
    let angle = phase * 2.0 * std::f64::consts::PI;
    (1.0 - angle.cos()) / 2.0 * 100.0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_duration() {
        assert_eq!(format_duration(30.0), "30s");
        assert_eq!(format_duration(90.0), "1m 30s");
        assert_eq!(format_duration(3661.0), "1h 1m 1s");
        assert_eq!(format_duration(90061.0), "1d 1h 1m 1s");
    }

    #[test]
    fn test_ra_conversion() {
        let (h, m, s) = decimal_to_ra(12.5);
        assert_eq!(h, 12);
        assert_eq!(m, 30);
        assert!((s - 0.0).abs() < 0.01);
        
        let decimal = ra_to_decimal(12, 30, 0.0);
        assert!((decimal - 12.5).abs() < 0.001);
    }

    #[test]
    fn test_dec_conversion() {
        let (d, m, s, neg) = decimal_to_dec(-45.5);
        assert_eq!(d, 45);
        assert_eq!(m, 30);
        assert!((s - 0.0).abs() < 0.01);
        assert!(neg);
        
        let decimal = dec_to_decimal(45, 30, 0.0, true);
        assert!((decimal + 45.5).abs() < 0.001);
    }
}
